# ТОВ «ТЕХНІЧНИЙ УНІВЕРСИТЕТ «МЕТІНВЕСТ ПОЛІТЕХНІКА»

## МЕТОДИЧНІ ВКАЗІВКИ ДО ВИКОНАННЯ ПРАКТИЧНОЇ РОБОТИ №4

### «Створення об'єктів, наслідування, поліморфізм»

**з навчальної дисципліни**

### «КОМП'ЮТЕРНА ТЕХНІКА, АЛГОРИТМІЗАЦІЯ ТА ПРОГРАМУВАННЯ»

---

**Запоріжжя 2025**



---

## ЗМІСТ

1. [Вступ](#вступ)
2. [Теоретичні відомості](#1-теоретичні-відомості)
3. [Хід роботи](#2-хід-роботи)
4. [Завдання](#3-завдання)
5. [Вимоги до оформлення звітів](#4-вимоги-до-оформлення-звітів)
6. [Додаток А. Варіанти завдань (1–23)](#додаток-а-варіанти-завдань)
7. [Додаток Б. Приклад виконання (Варіант 23)](#додаток-б-приклад-виконання-варіант-23)
8. [Додаток В. Типові помилки](#додаток-в-типові-помилки)
9. [Додаток Г. Глосарій](#додаток-г-глосарій)
10. [Список літератури](#список-літератури)

---

# ВСТУП

**Мета виконання практичної роботи:** набуття здобувачами навичок об'єктно-орієнтованого програмування (ООП) мовою Python: створення класів та об'єктів, організації наслідування, реалізації поліморфізму та застосування цих механізмів до розв'язання прикладної задачі.

**Завдання практичної роботи:** розробити структуру класів для обраної предметної області, реалізувати ієрархію з базовим класом та похідними класами, забезпечити поліморфну обробку об'єктів у колекції, виконати обчислення за формулами та подати розв'язок у вигляді консольного додатку.

**Час на виконання:** 6 академічних годин.

**Компетенції, що формуються:**
- Розуміння принципів об'єктно-орієнтованого програмування
- Вміння проектувати ієрархію класів
- Навички реалізації поліморфізму та наслідування
- Практичне застосування ООП до розв'язання прикладних задач

---

# 1 ТЕОРЕТИЧНІ ВІДОМОСТІ

## 1.1 Об'єктно-орієнтоване програмування: базові поняття

**Об'єктно-орієнтоване програмування (ООП)** — парадигма програмування, в якій програма описується як взаємодія об'єктів.

**Об'єкт** — це сутність, що має:
- **Стан** (дані, які зберігаються в атрибутах)
- **Поведінку** (операції, які реалізуються методами)

**Клас** — це шаблон (опис, blueprint) для створення об'єктів. Клас визначає, які атрибути та методи матимуть об'єкти.

**Об'єкт (екземпляр)** — це конкретна реалізація класу з конкретними значеннями атрибутів.

### Основні принципи ООП:

1. **Інкапсуляція** — об'єднання даних та методів, що працюють з цими даними, в межах класу
2. **Наслідування** — можливість створення нових класів на основі існуючих
3. **Поліморфізм** — можливість об'єктів різних класів реагувати на однакові виклики по-різному
4. **Абстракція** — виділення суттєвих характеристик об'єкта та ігнорування несуттєвих

### Приклад концепції:



In [15]:
# Клас — це шаблон
class Dog:
    def __init__(self, name, age):
        self.name = name  # атрибут
        self.age = age

    def __repr__(self):
        return f"Це собака на ім'я {self.name} віком {self.age} років"
    
    def bark(self):  # метод
        return f"{self.name} каже: Гав-гав!"

# Об'єкти — це конкретні екземпляри
dog1 = Dog("Рекс", 5)
dog2 = Dog("Бобік", 3)

print(dog1.bark())  # Рекс каже: Гав-гав!
print(dog2.bark())  # Бобік каже: Гав-гав!


Рекс каже: Гав-гав!
Бобік каже: Гав-гав!


In [14]:
print(dog1)

Це собака на ім'я Рекс віком 5



---

## 1.2 Створення класів та об'єктів у Python

### Оголошення класу

Клас у Python оголошується ключовим словом `class`. За конвенцією, назви класів пишуться в PascalCase.



In [36]:
class Example:
    """Документація класу (docstring)."""
    
    # Конструктор класу
    def __init__(self, a, b=0):
        """
        Ініціалізує об'єкт.
        
        Args:
            a: перший параметр
            b: другий параметр
        """
        self.a = a  # атрибут екземпляра
        self.b = b
    
    # Метод екземпляра
    def value(self):
        """Повертає суму a та b."""
        return self.a + self.b
    
    # Метод для виведення
    def __str__(self):
        return f"Example(a={self.a}, b={self.b})"

    def __repr__(self):
        return f"Instance of Example(a:{self.a}, b:{self.b})"


In [37]:
ex1 = Example(1, 2)

In [38]:
ex1.a

1

In [39]:
ex1.b

2

In [40]:
ex1.value()

3

In [41]:
ex1.value

<bound method Example.value of Instance of Example(a:1, b:2)>

In [42]:
str(ex1)

'Example(a=1, b=2)'

In [43]:
print(ex1)

Example(a=1, b=2)


In [44]:
ex1

Instance of Example(a:1, b:2)


### Створення об'єктів



In [45]:
# Створення екземпляра класу
obj = Example(10, 20)

# Виклик методів
result = obj.value()  # 30
print(obj)  # Example(a=10, b=20)

# Доступ до атрибутів
print(obj.a)  # 10


Example(a=10, b=20)
10



---

## 1.3 Наслідування

**Наслідування** — механізм створення нового класу (похідного, дочірнього) на основі існуючого (базового, батьківського).

### Переваги наслідування:
- Повторне використання коду
- Розширення функціональності
- Створення ієрархії класів

### Синтаксис наслідування:



In [91]:
class Base:
    """Базовий клас."""
    
    def __init__(self, x):
        self.x = x
    
    def calculate(self):
        return self.x * 2

    def power(self):
        return self.x ** self.x

class Child(Base):
    """Похідний клас."""
    
    def __init__(self, x, y):
        super().__init__(x)  # виклик конструктора базового класу
        self.y = y
    
    # Перевизначення методу (override)
    def calculate(self):
        return self.x * self.y
    
    # Новий метод
    def additional_method(self):
        return self.x + self.y

    def power(self):
        ans = super().power()
        print(f'{ans = }')
        res = ans // self.y
        print(f'{res = }')
        return res 


In [92]:
base1 = Base(10)

In [93]:
child1 = Child(20, 30)

In [94]:
base1.calculate()

20

In [95]:
child1.calculate()

600

In [96]:
child1.additional_method()

50

In [97]:
base1.additional_method()

AttributeError: 'Base' object has no attribute 'additional_method'

In [106]:
base1.power()

10000000000

In [107]:
child1.power()

ans = 104857600000000000000000000
res = 3495253333333333333333333


3495253333333333333333333


### Використання `super()`

`super()` дозволяє викликати методи базового класу з похідного:



In [103]:
class Animal:
    def __init__(self, name):
        self.name = name
    
    def speak(self):
        return "Звук тварини"

class Dog(Animal):
    def __init__(self, name, breed):
        super().__init__(name)  # виклик __init__ базового класу
        self.breed = breed
    
    def speak(self):
        super().speak()  # виклик методу базового класу
        return f"{self.name} каже: Гав!"


In [104]:
dog1 = Dog('Bob', 'вівчарка')

In [105]:
dog1.speak()

'Bob каже: Гав!'


---

## 1.4 Поліморфізм

**Поліморфізм** — можливість використовувати об'єкти різних класів через спільний інтерфейс, отримуючи різну поведінку залежно від фактичного типу об'єкта.

### Приклад поліморфізму:



In [109]:
x = 10
y = 20

In [111]:
print(type(x))

<class 'int'>


In [112]:
x + y

30

In [114]:
a = '10'
b = '20'

In [115]:
print(type(a))

<class 'str'>


In [116]:
a + b

'1020'

In [117]:
a.__add__(b)

'1020'

In [120]:
class StrInt(str):
    def __add__(self, b):
        ans = int(self) + int(b)
        return str(ans)

a = StrInt('10')
b = StrInt('20')
c = a + b

In [121]:
a

'10'

In [122]:
b

'20'

In [123]:
c

'30'

In [124]:
a * 3

'101010'

In [125]:
class Shape:
    def area(self):
        raise NotImplementedError("Метод має бути перевизначений")

class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius
    
    def area(self):
        return 3.14159 * self.radius ** 2

class Rectangle(Shape):
    def __init__(self, width, height):
        self.width = width
        self.height = height
    
    def area(self):
        return self.width * self.height

# Поліморфне використання
shapes = [Circle(5), Rectangle(4, 6), Circle(3)]

for shape in shapes:
    print(f"Площа: {shape.area():.2f}")
    # Викликається правильний метод залежно від типу об'єкта


Площа: 78.54
Площа: 24.00
Площа: 28.27



### Переваги поліморфізму:
- Гнучкість коду
- Легкість розширення
- Уніфікований інтерфейс

---

## 1.5 Інкапсуляція

**Інкапсуляція** — об'єднання даних та методів, що працюють з цими даними, у межах класу, з можливістю приховування внутрішньої реалізації.

### Рівні доступу в Python:



In [128]:
class Example:
    def __init__(self):
        self.public = "Публічний"        # доступний звідусіль
        self._protected = "Захищений"     # конвенція: для внутрішнього використання
        self.__private = "Приватний"      # name mangling

obj = Example()
print(obj.public)       # OK
print(obj._protected)   # OK, але не рекомендується
print(obj.__private)  # AttributeError


Публічний
Захищений


AttributeError: 'Example' object has no attribute '__private'

In [129]:
print(obj._Example__private)

Приватний



### Властивості (Properties):



In [130]:
class Temperature:
    def __init__(self, celsius):
        self._celsius = celsius
    
    @property
    def celsius(self):
        """Getter для температури."""
        return self._celsius
    
    @celsius.setter
    def celsius(self, value):
        """Setter з валідацією."""
        if value < -273.15:
            raise ValueError("Температура не може бути нижче абсолютного нуля")
        self._celsius = value
    
    @property
    def fahrenheit(self):
        """Обчислювана властивість."""
        return self._celsius * 9/5 + 32

temp = Temperature(25)
print(temp.celsius)      # 25
print(temp.fahrenheit)   # 77.0
temp.celsius = 30        # використання setter


25
77.0


In [138]:
temp.celsius = 300

In [139]:
temp.celsius

300


---

## 1.6 Абстрактні класи та інтерфейси

**Абстрактний клас** — клас, який не може бути інстанційований і призначений лише для наслідування.



In [9]:
from abc import ABC, abstractmethod

class AbstractShape(ABC):
    """Абстрактний базовий клас."""
    
    @abstractmethod
    def area(self):
        """Абстрактний метод — МАЄ бути перевизначений."""
        pass
    
    @abstractmethod
    def perimeter(self):
        """Ще один абстрактний метод."""
        pass
    
    def describe(self):
        """Конкретний метод (не абстрактний)."""
        return f"Фігура з площею {self.area()}"

class Square(AbstractShape):
    def __init__(self, side):
        self.side = side
    
    def area(self):
        return self.side ** 2
    
    def perimeter(self):
        return 4 * self.side

# shape = AbstractShape()  # TypeError: Can't instantiate abstract class
square = Square(5)  # OK



---

## 1.7 Магічні методи

**Магічні методи (dunder methods)** — спеціальні методи з подвійним підкресленням, які визначають поведінку об'єктів.

### Основні магічні методи:



In [10]:
class Point:
    def __init__(self, x, y):
        """Конструктор."""
        self.x = x
        self.y = y
    
    def __str__(self):
        """Рядкове представлення для користувача."""
        return f"Point({self.x}, {self.y})"
    
    def __repr__(self):
        """Рядкове представлення для розробника."""
        return f"Point(x={self.x}, y={self.y})"
    
    def __eq__(self, other):
        """Порівняння на рівність."""
        if not isinstance(other, Point):
            return False
        return self.x == other.x and self.y == other.y
    
    def __add__(self, other):
        """Додавання точок."""
        return Point(self.x + other.x, self.y + other.y)
    
    def __len__(self):
        """Довжина (відстань від початку координат)."""
        return int((self.x**2 + self.y**2)**0.5)

p1 = Point(3, 4)
p2 = Point(1, 2)

print(p1)           # Point(3, 4)
print(p1 + p2)      # Point(4, 6)
print(p1 == p2)     # False
print(len(p1))      # 5


Point(3, 4)
Point(4, 6)
False
5


In [140]:
list1 = [1, 2, 3, 4]

In [141]:
list1[1]

2

In [142]:
list1[1] = 22

In [143]:
list1

[1, 22, 3, 4]

In [144]:
list1.__getitem__(1)

22

In [146]:
list1.__setitem__(1, 222)

In [147]:
list1

[1, 222, 3, 4]


### Інші корисні магічні методи:

- `__lt__`, `__le__`, `__gt__`, `__ge__` — порівняння
- `__getitem__`, `__setitem__` — доступ через індекс
- `__call__` — виклик об'єкта як функції
- `__enter__`, `__exit__` — контекстний менеджер (with)

---

## 1.8 Формули та точність обчислень

У роботі необхідно використовувати числові обчислення з типом `float` та враховувати похибку округлення.

### Рекомендована константа точності:

$$\varepsilon = 10^{-9}$$



In [11]:
EPS = 1e-9



### Порівняння з точністю:

Замість:


In [12]:
x = .000_000_001
if x == 0:  # небезпечно для float
    ...



Використовуйте:


In [13]:
if abs(x) < EPS:  # безпечно
    ...



### Приклад порівняння інтервалів:



In [14]:
def in_range(value, min_val, max_val, eps=1e-9):
    """Перевіряє, чи значення в діапазоні з урахуванням похибки."""
    return (min_val - eps) <= value <= (max_val + eps)

# Використання
area = 0.0
s_min = 1.0
s_avg = 3.0
if in_range(area, s_min, s_avg):
    sum_interval += area



---

# 2 ХІД РОБОТИ

## Крок 1: Вивчення варіанту

1. Ознайомтеся з постановкою задачі вашого варіанту.
2. Виділіть сутності предметної області.
3. Визначте спільні та специфічні характеристики.

## Крок 2: Проектування ієрархії класів

1. Визначте базовий клас:
   - Які атрибути є спільними для всіх об'єктів?
   - Які методи повинні бути поліморфними?

2. Визначте похідні класи (2–3 класи):
   - Які специфічні атрибути потрібні?
   - Які методи потрібно перевизначити?



## Крок 3: Реалізація класів

Для кожного класу реалізуйте:
- Конструктор `__init__` з валідацією параметрів.
- Поліморфний метод обчислення (згідно варіанту).
- Метод `__str__` для форматованого виведення інформації про об'єкт.

## Крок 4: Головна програма

Реалізуйте функцію `main()`, що виконує:
1. Зчитування даних від користувача (кількість, параметри).
2. Створення колекції об'єктів (списку).
3. Поліморфну обробку (вивід таблиці).
4. Обчислення статистики (мін, макс, середнє, фільтрація).

## Крок 5: Тестування

Перевірте програму на тестових даних, наведених у вашому варіанті, а також протестуйте граничні випадки та обробку некоректного введення.

## Крок 6: Оформлення звіту

Оформіть звіт згідно з вимогами.

---

# 3 ЗАВДАННЯ

Розробити консольний додаток мовою Python, який реалізує об'єктно-орієнтовану модель (класи та об'єкти), наслідування та поліморфізм згідно з вашим варіантом.

## Обов'язкові вимоги:
- Реалізувати базовий та похідні класи.
- Використати поліморфізм.
- Реалізувати валідацію вхідних даних (додатність, діапазони тощо).
- Вивести результати у вигляді форматованої таблиці.
- Обчислити статистичні показники (сума, середнє, мін/макс).
- Код має відповідати стандарту PEP 8.

---

# 4 ВИМОГИ ДО ОФОРМЛЕННЯ ЗВІТІВ
Звіт має бути оформлений відповідно до вимог щодо оформлення технічної документації на аркушах формату А4 у друкованому вигляді.

## Структура звіту:

### 1. Титульна сторінка
- Назва закладу
- Назва роботи
- Варіант
- ПІБ студента, група
- ПІБ викладача
- Дата

### 2. Мета роботи
Стисло сформулювати мету (1-2 речення)

### 3. Постановка завдання
- Номер варіанту
- Повний текст завдання з додатку А
- Формули у математичному форматі
- Опис вхідних та вихідних даних

### 4. Реалізація



#### 5.1 Опис класів
Текстовий опис кожного класу:
- Призначення
- Атрибути
- Методи

#### 5.2 Лістинг програми
Повний код програми з коментарями

### 5. Тестування

#### 6.1 Тестові дані
Таблиця з вхідними даними

#### 6.2 Результати виконання
Скріншоти/текст виводу програми

#### 6.3 Перевірка коректності
Ручна перевірка обчислень для 1-2 об'єктів

### 6. Висновки
- Що було виконано
- Які навички набуто
- Які труднощі виникли
- Пропозиції щодо покращення

---



---

# ДОДАТОК А. ВАРІАНТИ ЗАВДАНЬ

**У кожному варіанті потрібно:**
1. Створити список об'єктів.
2. Вивести таблицю з результатами.
3. Знайти мінімальне та максимальне значення.
4. Обчислити середнє значення.
5. Обчислити загальний підсумок (суму) або виконати фільтрацію.

**Константа точності:** $\varepsilon = 10^{-9}$ (`EPS = 1e-9`)

---

## Варіант 1. "Геометричні фігури"

### Опис задачі:
Система обчислення площ геометричних фігур.

### Класи:
- **Базовий:** `Figure` (атрибут: `name`)
- **Похідні:** `Circle`, `Rectangle`, `Triangle`

### Поліморфний метод:
`area()` — обчислює площу фігури

### Формули:
**Коло:**
$$S_{circle}=\pi r^2$$

**Прямокутник:**
$$S_{rect}=a \cdot b$$

**Трикутник (за формулою Герона):**
$$S_{tri}=\sqrt{p(p-a)(p-b)(p-c)}, \quad p=\frac{a+b+c}{2}$$

де:
* $r$ — радіус кола;
* $a, b, c$ — сторони (прямокутника або трикутника);
* $p$ — півпериметр.

### Тестові дані:
- Коло: r = 5 → S ≈ 78.54
- Прямокутник: a = 4, b = 6 → S = 24
- Трикутник: a = 3, b = 4, c = 5 → S = 6

---

## Варіант 2. "Тарифи електроенергії"

### Опис задачі:
Система обчислення вартості електроенергії для різних типів споживачів.

### Класи:
- **Базовий:** `Tariff` (атрибути: `name`, `base_rate`)
- **Похідні:** `ResidentialTariff`, `CommercialTariff`, `IndustrialTariff`

### Поліморфний метод:
`cost(kwh)` — обчислює вартість за кількість спожитих кВт·год

### Формули:
**Житловий тариф:**
$$C_{res} = \begin{cases} kWh \cdot r_{base}, & kWh \leq 100 \\ 100 \cdot r_{base} + (kWh - 100) \cdot 1.5 \cdot r_{base}, & kWh > 100 \end{cases}$$

**Комерційний тариф:**
$$C_{com} = kWh \cdot 1.8 \cdot r_{base} + fee_{fixed}$$

**Промисловий тариф:**
$$C_{ind} = kWh \cdot 1.5 \cdot r_{base} \cdot (1 - discount)$$
де $discount = 0.1$ якщо $kWh > 1000$, інакше 0.

де:
* $kWh$ — спожита енергія (кВт·год);
* $r_{base}$ — базова ставка;
* $fee_{fixed}$ — фіксований збір.

### Тестові дані:
- Residential: 150 kWh, rate = 1.5 → C = 100·1.5 + 50·2.25 = 262.5
- Commercial: 200 kWh, rate = 1.5, fee = 50 → C = 200·2.7 + 50 = 590
- Industrial: 1500 kWh, rate = 1.5 → C = 1500·2.25·0.9 = 3037.5

---

## Варіант 3. "Доставка посилок"

### Опис задачі:
Система розрахунку вартості доставки для різних типів відправлень.

### Класи:
- **Базовий:** `Parcel` (атрибути: `weight`, `distance`)
- **Похідні:** `Box`, `Envelope`, `Pallet`

### Поліморфний метод:
`price()` — обчислює вартість доставки

### Формули:
**Коробка:**
$$P_{box} = 20 + w \cdot 5 + V \cdot 0.1 + d \cdot 0.5$$
де $V = length \cdot width \cdot height$ (об'єм), $w$ — вага (кг), $d$ — відстань (км).

**Конверт:**
$$P_{env} = 20 + w \cdot 5 + d \cdot 0.5 \cdot 0.5$$ (знижка на відстань 50%)

**Палета:**
$$P_{pal} = 20 + w \cdot 5 + V \cdot 0.1 \cdot 1.5 + d \cdot 0.5$$ (надбавка за об'єм 50%)

### Тестові дані:
- Box: w=5, V=30, d=100 → P = 20 + 25 + 3 + 50 = 98
- Envelope: w=0.5, d=50 → P = 20 + 2.5 + 12.5 = 35
- Pallet: w=100, V=500, d=200 → P = 20 + 500 + 75 + 100 = 695

---

## Варіант 4. "Транспортні квитки"

### Класи:
- **Базовий:** `Ticket` (атрибути: `distance`, `baggage`)
- **Похідні:** `BusTicket`, `TrainTicket`, `PlaneTicket`

### Поліморфний метод:
`total_price()`

### Формули:
* **Автобус:** $P = distance \cdot 0.5 + baggage \cdot 10$
* **Потяг:** $P = distance \cdot 1.2 \cdot class\_k + baggage \cdot 20$ (class_k: купе=1.5)
* **Літак:** $P = 500 + distance \cdot 2.0 \cdot class\_k + baggage \cdot 50$ (class_k: економ=1.0)

де:
* $distance$ — відстань (км);
* $baggage$ — кількість одиниць багажу;
* $class\_k$ — коефіцієнт класу обслуговування.

### Тестові дані:
- Bus: 100 км, 1 багаж → 60
- Train: 500 км, купе, 2 багажі → 940
- Plane: 1000 км, економ, 1 багаж → 2550

---

## Варіант 5. "Працівники та зарплата"

### Класи:
- **Базовий:** `Employee` (атрибути: `name`)
- **Похідні:** `HourlyEmployee` (hours, rate), `SalariedEmployee` (salary), `SalesEmployee` (base, sales, rate)

### Поліморфний метод:
`calculate_pay()`

### Формули:
* **Hourly:** $Pay = hours \cdot rate$ (якщо hours > 160, то надлишок $\cdot 1.5$)
* **Salaried:** $Pay = salary \cdot (1 + bonus\%)$
* **Sales:** $Pay = base + sales \cdot rate$

де:
* $hours$ — відпрацьовані години;
* $rate$ — погодинна ставка або відсоток від продажів;
* $salary$ — місячний оклад;
* $base$ — базова ставка (для відділу продажів);
* $sales$ — обсяг продажів;
* $bonus$ — відсоток премії.

### Тестові дані:
- Hourly: 180 год, 50/год → 8000 + 20·75 = 9500
- Salaried: 20000, +10% → 22000
- Sales: 10000 base, 100000 sales, 5% → 15000

---

## Варіант 6. "Кредити банку"

### Класи:
- **Базовий:** `Loan` (sum, rate, term_months)
- **Похідні:** `AnnuityLoan`, `DifferentiatedLoan`

### Поліморфний метод:
`total_overpayment()` — загальна переплата за кредитом

### Формули:
**Annuity:**
$MonthPay = S \cdot \frac{r(1+r)^n}{(1+r)^n - 1}$, де $r = rate/12/100$
$Overpayment = MonthPay \cdot n - S$

**Differentiated:**
$MonthPay_k = \frac{S}{n} + (S - \frac{S}{n}(k-1)) \cdot r$
$TotalPay = \sum MonthPay_k$
$Overpayment = TotalPay - S$

де:
* $S$ — початкова сума кредиту;
* $r$ — місячна відсоткова ставка (річна / 12 / 100);
* $n$ — термін кредиту (місяців);
* $k$ — номер поточного місяця.

### Тестові дані (100000, 12%, 12 міс):
- Annuity Overpayment: ~6618.55
- Differentiated Overpayment: ~6500.00

---

## Варіант 7. "Податки на доходи"

### Класи:
- **Базовий:** `TaxPayer` (income)
- **Похідні:** `Individual`, `Entrepreneur` (expenses), `Company` (expenses)

### Поліморфний метод:
`calculate_tax()`

### Формули:
* **Individual:** 18% (до 300к), 25% (з суми перевищення 300к) + 54000
* **Entrepreneur:** $(income \cdot 0.05) - (expenses \cdot 0.7)$ (не менше 0)
* **Company:** $(income - expenses) \cdot 0.18$

де:
* $income$ — дохід;
* $expenses$ — документально підтверджені витрати.

### Тестові дані:
- Indiv (400k) → 54k + 25k = 79000
- Entr (500k, exp 200k) → 25k - 14k = 11000 (якщо <0 то 0)
- Comp (1M, exp 600k) → 400k · 0.18 = 72000

---

## Варіант 8. "Паливо та витрати авто"

### Класи:
- **Базовий:** `Vehicle` (consumption_l_100km)
- **Похідні:** `Car`, `Truck`, `Motorcycle`

### Поліморфний метод:
`trip_cost(distance, fuel_price)`

### Формули:
* **Car:** $Cost = \frac{dist}{100} \cdot cons \cdot price + 100$ (toll)
* **Truck:** $Cost = \frac{dist}{100} \cdot cons \cdot 1.3 \cdot price + 200$ (toll)
* **Moto:** $Cost = \frac{dist}{100} \cdot cons \cdot 0.8 \cdot price$ (no toll)

де:
* $dist$ — відстань поїздки (км);
* $cons$ — витрата палива (л/100 км);
* $price$ — ціна палива за літр.

### Тестові дані (500 км, 50 грн/л):
- Car (8л): 40л · 50 + 100 = 2100
- Truck (25л): 162.5л · 50 + 200 = 8325
- Moto (5л): 20л · 50 = 1000

---

## Варіант 9. "Страхові поліси"

### Опис задачі:
Система розрахунку вартості страхових премій.

### Класи:
- **Базовий:** `Insurance` (атрибут: `coverage_amount`)
- **Похідні:** `HealthInsurance`, `CarInsurance`, `PropertyInsurance`

### Поліморфний метод:
`premium()` — обчислює вартість полісу

### Формули:
**Медичне страхування (Health):**
$$P = coverage \cdot 0.05 \cdot age\_factor$$
де $age\_factor$: 0.8 (<30 років), 1.0 (30-50 років), 1.5 (>50 років)

**Автострахування (Car):**
$$P = car\_value \cdot 0.04 \cdot risk$$
де $risk$: 1.0 (0 ДТП), 1.2 (1 ДТП), 1.5 (>1 ДТП)

**Майно (Property):**
$$P = value \cdot 0.002 \cdot location$$
де $location$: 1.0 (місто), 0.8 (село), 1.3 (центр)

де:
* $coverage$ — сума страхового покриття;
* $age\_factor$ — віковий коефіцієнт;
* $car\_value$ — вартість автомобіля;
* $risk$ — коефіцієнт ризику (історія ДТП);
* $value$ — оціночна вартість майна;
* $location$ — коефіцієнт розташування.

### Тестові дані:
- Health: 100000, вік 35 → 5000
- Car: 200000, 1 ДТП → 9600
- Property: 1000000, центр → 2600

---

## Варіант 10. "Знижки магазину"

### Опис задачі:
Розрахунок фінальної ціни товару з урахуванням різних типів знижок.

### Класи:
- **Базовий:** `Discount` (атрибут: `original_price`)
- **Похідні:** `PercentDiscount`, `FixedDiscount`, `ThresholdDiscount`

### Поліморфний метод:
`final_price()` — повертає ціну зі знижкою

### Формули:
**Відсоткова (Percent):**
$$P_{final} = P \cdot (1 - rate)$$

**Фіксована (Fixed):**
$$P_{final} = \max(0, P - amount)$$

**Порогова (Threshold):**
$$P_{final} = \begin{cases} P \cdot (1 - rate), & P > threshold \\ P, & P \leq threshold \end{cases}$$

де:
* $P$ — початкова ціна товару;
* $rate$ — розмір знижки (у десятковому дробі, напр. 0.1 = 10%);
* $amount$ — фіксована сума знижки;
* $threshold$ — поріг ціни для застосування знижки.

### Тестові дані:
- Percent: 1000, 10% → 900
- Fixed: 500, знижка 100 → 400
- Threshold: 1000, поріг 800, 20% → 800

---

## Варіант 11. "Навчальні курси"

### Опис задачі:
Розрахунок вартості навчання на різних типах курсів.

### Класи:
- **Базовий:** `Course` (атрибут: `hours`)
- **Похідні:** `OnlineCourse`, `OfflineCourse`, `HybridCourse`

### Поліморфний метод:
`total_cost()`

### Формули:
**Online:**
$$Cost = 500 + hours \cdot 50$$

**Offline:**
$$Cost = 1000 + hours \cdot 100 + materials$$

**Hybrid:**
$$Cost = 750 + hours \cdot 80 + materials \cdot 0.5$$

де:
* $hours$ — тривалість курсу (год);
* $materials$ — вартість навчальних матеріалів.

### Тестові дані:
- Online: 50 год → 3000
- Offline: 20 год, мат. 500 → 3500
- Hybrid: 30 год, мат. 500 → 3400

---

## Варіант 12. "Фізична енергія"

### Опис задачі:
Обчислення енергії різних фізичних систем.

### Класи:
- **Базовий:** `EnergySystem` (атрибут: `mass`)
- **Похідні:** `KineticEnergy`, `PotentialEnergy`, `ThermalEnergy`

### Поліморфний метод:
`calculate_energy()`

### Формули:
**Кінетична (Kinetic):**
$$E_k = \frac{m \cdot v^2}{2}$$

**Потенціальна (Potential):**
$$E_p = m \cdot g \cdot h \quad (g=9.81)$$

**Теплова (Thermal, нагрівання):**
$$E_t = m \cdot c \cdot \Delta T$$

де:
* $m$ — маса (кг);
* $v$ — швидкість (м/с);
* $h$ — висота (м);
* $c$ — питома теплоємність;
* $\Delta T$ — зміна температури.

### Тестові дані:
- Kinetic: m=5, v=10 → 250
- Potential: m=10, h=5 → 490.5
- Thermal: m=2, c=4200 (вода), dt=10 → 84000

---

## Варіант 13. "Конвертація валют"

### Опис задачі:
Розрахунок суми до отримання при обміні валют різними способами.

### Класи:
- **Базовий:** `CurrencyExchange` (атрибут: `amount_usd`)
- **Похідні:** `CashExchange`, `CardExchange`, `CryptoExchange`

### Поліморфний метод:
`get_uah_amount(rate)`

### Формули:
**Готівка (Cash):**
$$Sum = amount \cdot rate - fee\_fixed$$ (fee=20 грн)

**Картка (Card):**
$$Sum = amount \cdot rate \cdot (1 - 0.01)$$ (1% комісія)

**Криптовалюта (Crypto):**
$$Sum = amount \cdot rate \cdot (1 - 0.02) + bonus$$ (бонус 50 грн, комісія 2%)

де:
* $amount$ — сума у валюті (USD);
* $rate$ — поточний курс обміну;
* $fee\_fixed$ — фіксована комісія (20 грн);
* $bonus$ — додатковий бонус.

### Тестові дані (100 USD, курс 40):
- Cash: 4000 - 20 = 3980
- Card: 4000 - 40 = 3960
- Crypto: 4000 - 80 + 50 = 3970

---

## Варіант 14. "Об'єм тіл обертання"

### Опис задачі:
Обчислення об'ємів геометричних тіл.

### Класи:
- **Базовий:** `Solid3D` (атрибут: `radius`)
- **Похідні:** `Sphere`, `Cylinder`, `Cone`

### Поліморфний метод:
`volume()`

### Формули:
**Сфера (Sphere):**
$$V = \frac{4}{3} \pi r^3$$

**Циліндр (Cylinder):**
$$V = \pi r^2 h$$

**Конус (Cone):**
$$V = \frac{1}{3} \pi r^2 h$$

де:
* $r$ — радіус основи;
* $h$ — висота тіла.

### Тестові дані (r=3, h=10):
- Sphere: ~113.097
- Cylinder: ~282.743
- Cone: ~94.248

---

## Варіант 15. "Амортизація активів"

### Опис задачі:
Розрахунок річної амортизації основних засобів різними методами.

### Класи:
- **Базовий:** `Asset` (cost, salvage_value, life_years)
- **Похідні:** `StraightLine`, `SumOfYears`, `DoubleDeclining`

### Поліморфний метод:
`depreciation(year_number)`

### Формули:
**Прямолінійний (StraightLine):**
$$D = \frac{Cost - Salvage}{Life}$$

**Сума років (SumOfYears):**
$$D = (Cost - Salvage) \cdot \frac{Life - Year + 1}{\sum years}$$

**Подвійний залишок (DoubleDeclining):**
$$D = (Cost - AccumulatedDep) \cdot \frac{2}{Life}$$

де:
* $Cost$ — початкова вартість активу;
* $Salvage$ — ліквідаційна вартість;
* $Life$ — термін корисного використання (років);
* $Year$ — номер року експлуатації;
* $AccumulatedDep$ — накопичена амортизація за попередні періоди.

### Тестові дані (Cost=10000, Salvage=1000, Life=5, Year=1):
- Straight: (9000)/5 = 1800
- SumYears: 9000 * 5/15 = 3000
- Double: 10000 * 0.4 = 4000

---

## Варіант 16. "Квитки в музей"

### Опис задачі:
Розрахунок ціни квитка для різних категорій відвідувачів.

### Класи:
- **Базовий:** `Visitor` (base_price)
- **Похідні:** `Child`, `Student`, `Adult`, `Senior`

### Поліморфний метод:
`ticket_price()`

### Формули:
* **Child (<6 років):** Безкоштовно (0)
* **Student:** 50% знижки ($P \cdot 0.5$)
* **Adult:** Повна ціна ($P$)
* **Senior (>60 років):** 30% знижки ($P \cdot 0.7$)

де:
* $P$ — базова ціна квитка.

### Тестові дані (Base=200):
- Child: 0
- Student: 100
- Adult: 200
- Senior: 140

---

## Варіант 17. "Оренда житла"

### Опис задачі:
Розрахунок вартості оренди залежно від терміну та типу.

### Класи:
- **Базовий:** `Rental` (rate_per_day)
- **Похідні:** `DailyRent`, `WeeklyRent`, `MonthlyRent`

### Поліморфний метод:
`calculate_rent(days)`

### Формули:
* **Подобова:** $Cost = days \cdot rate$
* **Тижнева:** $Cost = days \cdot rate \cdot 0.85$ (знижка 15% при оренді > 7 днів)
* **Місячна:** $Cost = days \cdot rate \cdot 0.60$ (знижка 40% при оренді > 30 днів)

де:
* $days$ — кількість днів оренди;
* $rate$ — базова добова вартість.

### Тестові дані (Rate=1000, Days=35):
- Daily: 35000
- Weekly: 35000 * 0.85 = 29750
- Monthly: 35000 * 0.6 = 21000

---

## Варіант 18. "Датчики температури"

### Опис задачі:
Обробка показань різних типів датчиків температури (калібрування).

### Класи:
- **Базовий:** `Sensor` (raw_value)
- **Похідні:** `AnalogSensor`, `DigitalSensor`, `SmartSensor`

### Поліморфний метод:
`get_temperature()`

### Формули:
* **Аналоговий:** $T = raw \cdot 0.1 - 5$ (лінійне зміщення)
* **Цифровий:** $T = raw \cdot 0.05$ (просте масштабування)
* **Розумний:** $T = raw \cdot 0.05 + correction$ (де correction залежить від діапазону: якщо raw>500, то -2, інакше +1)

де:
* $T$ — результуюча температура;
* $raw$ — "сирі" дані з датчика;
* $correction$ — коригувальне значення.

### Тестові дані (Raw=600):
- Analog: 60 - 5 = 55.0
- Digital: 30.0
- Smart: 30 - 2 = 28.0

---

## Варіант 19. "Банківські вклади"

### Опис задачі:
Прогнозування суми на депозиті.

### Класи:
- **Базовий:** `Deposit` (amount, rate, years)
- **Похідні:** `SimpleDeposit`, `CompoundDeposit`, `BonusDeposit`

### Поліморфний метод:
`final_amount()`

### Формули:
**Простий:**
$$S = P \cdot (1 + r \cdot t)$$

**Складний (капіталізація щороку):**
$$S = P \cdot (1 + r)^t$$

**Бонусний:**
$$S = P \cdot (1 + r \cdot t) + bonus$$ (bonus = 2% від P, якщо t > 3)

де:
* $P$ — сума вкладу (Principial);
* $r$ — річна відсоткова ставка (у десятковому дробі);
* $t$ — термін вкладу (років);
* $bonus$ — додатковий бонус.

### Тестові дані (10000, 10% (0.1), 4 роки):
- Simple: 10000 * 1.4 = 14000
- Compound: 10000 * 1.4641 = 14641
- Bonus: 14000 + 200 = 14200

---

## Варіант 20. "Мобільний зв'язок"

### Опис задачі:
Розрахунок вартості послуг мобільного зв'язку.

### Класи:
- **Базовий:** `MobilePlan` (minutes, gb)
- **Похідні:** `Prepaid`, `Postpaid`, `Unlimited`

### Поліморфний метод:
`total_bill()`

### Формули:
* **Prepaid:** $Min \cdot 1.0 + GB \cdot 50$
* **Postpaid:** $150 (base) + (Min-100)_+ \cdot 0.5 + (GB-5)_+ \cdot 30$
* **Unlimited:** Fixed 350

де:
* $Min$ — використані хвилини;
* $GB$ — обсяг використаного трафіку (Гб).

### Тестові дані (200 хв, 10 Гб):
- Prepaid: 200 + 500 = 700
- Postpaid: 150 + 100*0.5 + 5*30 = 150 + 50 + 150 = 350
- Unlimited: 350

---

## Варіант 21. "Інтернет-магазин: кошик"

### Опис задачі:
Розрахунок вартості товарів у кошику з урахуванням категорії покупця.

### Класи:
- **Базовий:** `Buyer` (cart_total)
- **Похідні:** `NewBuyer`, `RegularBuyer`, `VipBuyer`

### Поліморфний метод:
`checkout_price()`

### Формули:
* **Новий:** Без знижки ($Total$)
* **Постійний:** 5% знижки ($Total \cdot 0.95$)
* **VIP:** 10% знижки + якщо сума > 1000, ще -50 грн ($Total \cdot 0.90 - 50$)

де:
* $Total$ — загальна вартість товарів у кошику без знижок.

### Тестові дані (1500 грн):
- New: 1500
- Regular: 1425
- VIP: 1500 * 0.9 - 50 = 1350 - 50 = 1300

---

## Варіант 22. "Служба таксі"

### Опис задачі:
Розрахунок вартості поїздки різними класами авто.

### Класи:
- **Базовий:** `Taxi` (distance_km)
- **Похідні:** `Economy`, `Comfort`, `Business`

### Поліморфний метод:
`fare()`

### Формули:
* **Economy:** $35 + dist \cdot 10$
* **Comfort:** $50 + dist \cdot 12 + 10$ (кондиціонер)
* **Business:** $80 + dist \cdot 20 + 50$ (сервіс)

де:
* $dist$ — відстань поїздки (км).

### Тестові дані (10 км):
- Economy: 35 + 100 = 135
- Comfort: 50 + 120 + 10 = 180
- Business: 80 + 200 + 50 = 330

---

## Варіант 23. "Площі фігур та статистика"

### Опис задачі:
Система обчислення площ різних геометричних фігур та збору статистики.

### Класи:
- **Базовий:** `Shape` (атрибут: `name`)
- **Похідні:** `Square`, `Circle`, `RightTriangle`

### Поліморфний метод:
`area()` — обчислює площу

### Формули:
**Квадрат:**
$$S_{square} = a^2$$

**Коло:**
$$S_{circle} = \pi r^2$$

**Прямокутний трикутник:**
$$S_{triangle} = \frac{1}{2} a \cdot b$$

де:
* $a$ — сторона квадрата або перший катет;
* $b$ — другий катет;
* $r$ — радіус кола.

### Завдання:
1. Створити список фігур.
2. Вивести детальну таблицю.
3. Знайти мін, макс, середню площу.
4. Знайти суму площ, що потрапляють в інтервал $[S_{min}; S_{avg}]$.

### Тестові дані:
- Square a=2 → 4
- Circle r=1 → 3.14159
- Triangle a=3, b=4 → 6

---
# ДОДАТОК Б. ПРИКЛАД ВИКОНАННЯ (ВАРІАНТ 23)

## Повний лістинг програми



In [21]:
import math

EPS = 1e-9


class Shape:
    """Базовий клас для геометричних фігур."""
    
    def __init__(self, name):
        self.name = name
    
    def area(self):
        """Обчислює площу фігури."""
        raise NotImplementedError("Метод має бути перевизначений")
    
    def params_str(self):
        """Повертає рядок з параметрами фігури."""
        return "-"
    
    def __str__(self):
        """Рядкове представлення."""
        return f"{self.name}: {self.params_str()}, площа={self.area():.6f}"


class Square(Shape):
    """Клас для квадрата."""
    
    def __init__(self, a):
        super().__init__("Квадрат")
        if a <= 0:
            raise ValueError("Сторона має бути додатною")
        self.a = a
    
    def area(self):
        return self.a * self.a
    
    def params_str(self):
        return f"a={self.a:.6f}"


class Circle(Shape):
    """Клас для кола."""
    
    def __init__(self, r):
        super().__init__("Коло")
        if r <= 0:
            raise ValueError("Радіус має бути додатним")
        self.r = r
    
    def area(self):
        return math.pi * self.r * self.r
    
    def params_str(self):
        return f"r={self.r:.6f}"


class RightTriangle(Shape):
    """Клас для прямокутного трикутника."""
    
    def __init__(self, a, b):
        super().__init__("Прямокутний трикутник")
        if a <= 0 or b <= 0:
            raise ValueError("Катети мають бути додатними")
        self.a = a
        self.b = b
    
    def area(self):
        return (self.a * self.b) / 2.0
    
    def params_str(self):
        return f"a={self.a:.6f}, b={self.b:.6f}"


def read_positive_float(prompt):
    """Зчитує додатне дійсне число з валідацією."""
    while True:
        try:
            x = float(input(prompt))
            if x <= 0:
                print("❌ Помилка: значення має бути додатним")
                continue
            return x
        except ValueError:
            print("❌ Помилка: потрібно ввести число")


def read_int_in_range(prompt, lo, hi):
    """Зчитує ціле число в діапазоні з валідацією."""
    while True:
        try:
            n = int(input(prompt))
            if n < lo or n > hi:
                print(f"❌ Помилка: число має бути в діапазоні [{lo}; {hi}]")
                continue
            return n
        except ValueError:
            print("❌ Помилка: потрібно ввести ціле число")


def create_shape(index):
    """Створює фігуру на основі введення користувача."""
    print(f"\n{'─' * 70}")
    print(f"Фігура #{index}")
    print("Оберіть тип:")
    print("  1 — Квадрат")
    print("  2 — Коло")
    print("  3 — Прямокутний трикутник")
    
    shape_type = read_int_in_range("Тип (1/2/3): ", 1, 3)
    
    try:
        if shape_type == 1:
            a = read_positive_float("Введіть сторону a: ")
            return Square(a)
        elif shape_type == 2:
            r = read_positive_float("Введіть радіус r: ")
            return Circle(r)
        else:
            a = read_positive_float("Введіть катет a: ")
            b = read_positive_float("Введіть катет b: ")
            return RightTriangle(a, b)
    except ValueError as e:
        print(f"❌ Помилка створення фігури: {e}")
        return None


def main():
    """Головна функція програми."""
    print("=" * 70)
    print("ПРАКТИЧНА РОБОТА №4: ООП")
    print("Варіант 23: Площі фігур та підсумкова статистика")
    print("=" * 70)
    
    # Введення кількості фігур
    n = read_int_in_range("\nВведіть кількість фігур N (1-100): ", 1, 100)
    
    # Створення колекції фігур
    shapes = []
    for i in range(1, n + 1):
        shape = create_shape(i)
        if shape:
            shapes.append(shape)
    
    if not shapes:
        print("\n❌ Не створено жодної фігури!")
        return
    
    # Обчислення площ
    print("\n" + "=" * 70)
    print(f"{'№':>5} | {'Тип':<26} | {'Параметри':<28} | {'Площа':>12}")
    print("─" * 70)
    
    areas = []
    for idx, shape in enumerate(shapes, start=1):
        s = shape.area()
        areas.append(s)
        print(f"{idx:5d} | {shape.name:<26} | {shape.params_str():<28} | {s:12.6f}")
    
    # Статистика
    s_min = min(areas)
    s_max = max(areas)
    s_sum = sum(areas)
    s_avg = s_sum / len(areas)
    
    # Сума в інтервалі [S_min; S_avg]
    sum_interval = 0.0
    count_interval = 0
    for s in areas:
        if (s_min - EPS) <= s <= (s_avg + EPS):
            sum_interval += s
            count_interval += 1
    
    # Виведення результатів
    print("=" * 70)
    print("СТАТИСТИКА:")
    print("─" * 70)
    print(f"S_min (мінімальна площа) = {s_min:.6f}")
    print(f"S_max (максимальна площа) = {s_max:.6f}")
    print(f"S_avg (середня площа)    = {s_avg:.6f}")
    print(f"S_sum (сума всіх площ)   = {s_sum:.6f}")
    print("─" * 70)
    print(f"Сума площ у інтервалі [S_min; S_avg] = {sum_interval:.6f}")
    print(f"Кількість фігур у інтервалі          = {count_interval}")
    print("=" * 70)


if __name__ == "__main__":
    main()


ПРАКТИЧНА РОБОТА №4: ООП
Варіант 23: Площі фігур та підсумкова статистика



Введіть кількість фігур N (1-100):  3



──────────────────────────────────────────────────────────────────────
Фігура #1
Оберіть тип:
  1 — Квадрат
  2 — Коло
  3 — Прямокутний трикутник


Тип (1/2/3):  1
Введіть сторону a:  2



──────────────────────────────────────────────────────────────────────
Фігура #2
Оберіть тип:
  1 — Квадрат
  2 — Коло
  3 — Прямокутний трикутник


Тип (1/2/3):  2
Введіть радіус r:  1



──────────────────────────────────────────────────────────────────────
Фігура #3
Оберіть тип:
  1 — Квадрат
  2 — Коло
  3 — Прямокутний трикутник


Тип (1/2/3):  3
Введіть катет a:  3
Введіть катет b:  4



    № | Тип                        | Параметри                    |        Площа
──────────────────────────────────────────────────────────────────────
    1 | Квадрат                    | a=2.000000                   |     4.000000
    2 | Коло                       | r=1.000000                   |     3.141593
    3 | Прямокутний трикутник      | a=3.000000, b=4.000000       |     6.000000
СТАТИСТИКА:
──────────────────────────────────────────────────────────────────────
S_min (мінімальна площа) = 3.141593
S_max (максимальна площа) = 6.000000
S_avg (середня площа)    = 4.380531
S_sum (сума всіх площ)   = 13.141593
──────────────────────────────────────────────────────────────────────
Сума площ у інтервалі [S_min; S_avg] = 7.141593
Кількість фігур у інтервалі          = 2


# ДОДАТОК Г. ТИПОВІ ПОМИЛКИ

## 1. Відсутність виклику super().__init__()

❌ **Неправильно:**

In [16]:
class Circle(Shape):
    def __init__(self, r):
        self.r = r  # не викликано super()



✅ **Правильно:**


In [17]:
class Circle(Shape):
    def __init__(self, r):
        super().__init__("Коло")
        self.r = r



## 2. Відсутність валідації

❌ **Неправильно:**


In [18]:
def __init__(self, a):
    self.a = a  # може бути від'ємним або нулем!



✅ **Правильно:**


In [19]:
def __init__(self, a):
    if a <= 0:
        raise ValueError("Сторона має бути додатною")
    self.a = a



## 3. Порівняння float без точності

❌ **Неправильно:**


In [22]:
if area == s_avg:  # небезпечно для float
    ...



✅ **Правильно:**


In [23]:
if abs(area - s_avg) < EPS:
    ...



або для інтервалів:


In [24]:
if (s_min - EPS) <= area <= (s_avg + EPS):
    ...



## 4. Відсутність обробки винятків

❌ **Неправильно:**


In [25]:
x = float(input("Число: "))  # може бути ValueError


Число:  1



✅ **Правильно:**


In [ ]:
try:
    x = float(input("Число: "))
except ValueError:
    print("Помилка: введіть число")



## 5. Відсутність docstrings

❌ **Неправильно:**


In [ ]:
class Shape:
    def area(self):
        pass



✅ **Правильно:**


In [ ]:
class Shape:
    """Базовий клас для геометричних фігур."""
    
    def area(self):
        """Обчислює площу фігури.
        
        Returns:
            float: Площа фігури
        """
        pass



## 6. Прямий доступ до атрибутів без перевірки

❌ **Неправильно:**


In [ ]:
shape.name = ""  # пряма зміна



✅ **Правильно (якщо потрібна валідація):**


In [ ]:
@property
def name(self):
    return self._name

@name.setter
def name(self, value):
    if not value:
        raise ValueError("Ім'я не може бути порожнім")
    self._name = value



## 7. Неефективна реалізація

❌ **Неправильно:**


In [ ]:
# Обчислення площі кілька разів
if shape.area() > 10:
    print(shape.area())
    sum += shape.area()



✅ **Правильно:**


In [ ]:
area = shape.area()  # обчислити один раз
if area > 10:
    print(area)
    sum += area



---

# ДОДАТОК Д. ГЛОСАРІЙ

**Абстрактний клас** — клас, який не може бути інстанційований і призначений лише для наслідування.

**Атрибут** — змінна, що належить об'єкту або класу.

**Базовий клас (батьківський)** — клас, від якого успадковуються інші класи.

**Інкапсуляція** — об'єднання даних та методів в межах класу з можливістю приховування реалізації.

**Клас** — шаблон для створення об'єктів, що визначає їх структуру та поведінку.

**Конструктор** — метод `__init__`, що викликається при створенні об'єкта.

**Магічні методи (dunder)** — спеціальні методи з подвійним підкресленням (\_\_init\_\_, \_\_str\_\_).

**Метод** — функція, що належить класу та визначає поведінку об'єктів.

**Множинне наслідування** — наслідування від кількох батьківських класів одночасно.

**Наслідування** — механізм створення нового класу на основі існуючого.

**Об'єкт (екземпляр)** — конкретна реалізація класу з конкретними значеннями атрибутів.

**ООП** — Об'єктно-орієнтоване програмування.

**Перевизначення (override)** — заміна методу батьківського класу в похідному класі.

**Поліморфізм** — можливість об'єктів різних класів реагувати на однакові виклики по-різному.

**Похідний клас (дочірній)** — клас, створений шляхом наслідування від базового класу.

**Self** — посилання на поточний об'єкт всередині методів класу.

**Super()** — функція для виклику методів батьківського класу.

---

# СПИСОК ЛІТЕРАТУРИ

1. **Lutz M.** Learning Python, 5th Edition. — O'Reilly Media, 2013. — 1600 p.

2. **Summerfield M.** Programming in Python 3: A Complete Introduction to the Python Language. — Addison-Wesley Professional, 2009.

3. **Reitz K., Schlusser T.** The Hitchhiker's Guide to Python: Best Practices for Development. — O'Reilly Media, 2016.

4. **Офіційна документація Python:** https://docs.python.org/3/

5. **PEP 8 — Style Guide for Python Code:** https://www.python.org/dev/peps/pep-0008/

6. **Real Python Tutorials:** https://realpython.com/

7. **Python OOP Tutorial (Corey Schafer):** https://www.youtube.com/watch?v=ZDa-Z5JzLYM

8. **Proglib - уроки ООП на Python:** https://proglib.io/p/python-oop

9. **Метод Герона для обчислення площі трикутника:** https://uk.wikipedia.org/wiki/Формула_Герона

10. **Ануїтетний та диференційований платіж:** НБУ, Методичні рекомендації з кредитування.

---

**© ТОВ «Технічний університет «Метінвест Політехніка», 2025**

